cellrank/tl/_utils.py

```diff
@@ -321,39 +332,224 @@ def _bias_knn(
+    corr, pvals, ci_low, ci_high = _correlation_test_helper(
+        X.T, Y.X, method=method, n_perms=n_perms, seed=seed, **kwargs
+    )
+    res = pd.DataFrame(corr, index=gene_names, columns=[f"{c} corr" for c in Y.names])
+    res[[f"{c} pval" for c in Y.names]] = pvals
+    res[[f"{c} qval" for c in Y.names]] = res[[f"{c} pval" for c in Y.names]].apply(
+        lambda c: multipletests(c, alpha=0.05, method="fdr_bh")[1]
     )
+    res[[f"{c} ci low" for c in Y.names]] = ci_low
+    res[[f"{c} ci high" for c in Y.names]] = ci_high
+
+    res = res[
+        [
+            f"{c} {stat}"
+            for c in Y.names
+            for stat in ("corr", "pval", "qval", "ci low", "ci high")
+        ]
+    ]
+    res.sort_values(by=[f"{c} corr" for c in Y.names], ascending=False, inplace=True)
 
-    if np.sum(denom == 0):
-        logg.warning(
-            f"No variation found in `{np.sum(denom==0)}` genes. Correlation for these will be `NaN`"
+    return res
+
+
+def _correlation_test_helper(
+    X: Union[np.ndarray, spmatrix],
+    Y: np.ndarray,
+    method: TestMethod = TestMethod.FISCHER,
+    n_perms: Optional[int] = None,
+    seed: Optional[int] = None,
+    **kwargs,
+) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
+    """
+    Compute the correlation between rows in matrix ``X`` columns of matrix ``Y``.
+
+    Parameters
+    ----------
+    X
+        Array or matrix of `(M, N)` elements.
+    Y
+        Array of `(N, K)` elements.
+    method
+        Method for p-value calculation.
+    n_perms
+        Number of permutations if ``method='perm_test'``.
+    seed
+        Random seed if ``method='perm_test'``.
+    **kwargs
+        Keyword arguments for :func:`cellrank.ul._parallelize.parallelize`.
+
+    Returns
+    -------
+        Correlations, p-values, corrected p-values, lower and upper bound of 95% confidence interval.
+        Each array if of shape ``(n_genes, n_lineages)``.
+    """
+
+    def perm_test_extractor(
+        res: Sequence[Tuple[np.ndarray, np.ndarray]]
+    ) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
+        pvals, corr_bs = zip(*res)
+        pvals = np.sum(pvals, axis=0) / float(n_perms)
+
+        corr_bs = np.concatenate(corr_bs, axis=0)
+        corr_ci_low, corr_ci_high = np.quantile(corr_bs, q=0.025, axis=0), np.quantile(
+            corr_bs, q=0.975, axis=0
         )
 
-    return num / denom
+        return pvals, corr_ci_low, corr_ci_high
+
+    n = X.shape[1]  # genes x cells
+
+    if issparse(X) and not isspmatrix_csr(X):
+        X = csr_matrix(X)
+
+    corr = _mat_mat_corr_sparse(X, Y) if issparse(X) else _mat_mat_corr_dense(X, Y)
+
+    if method == TestMethod.FISCHER:
+        mean, se = np.arctanh(corr), 1.0 / np.sqrt(n - 3)
+        z_score = (np.arctanh(corr) - np.arctanh(0)) * np.sqrt(n - 3)
+
+        z = norm.ppf(0.975)
+        corr_ci_low = np.tanh(mean - z * se)
+        corr_ci_high = np.tanh(mean + z * se)
+        pvals = 2 * norm.cdf(-np.abs(z_score))
```

Comment:
The maths seems correct, but would greatly benefit from some comments so we can still understand what we did in 4 weeks.

Reply:
e.g. why is z defined the way it is?

Reply:
This clearly depends on the size of the confidence interval (which should, by the way, probably be a method parameter)

Reply:
Ok, I will expose the width of the CI (didn't want to overwhelm the user, but it might be useful).